### PR TITLE
Various fixes to the lazy core

### PR DESCRIPTION
**Pull request by asuhan**
Discovered a handful of issues during the XLA plug-in bring up, fixed them.

### CODE REVIEW

**Review by wconstab — COMMENTED** [review]
mostly lgtm.  did you have a way to test any of the changes?  esp. the hash lib?

File: lazy_tensor_core/third_party/computation_client/ltc_util.cc [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  for (auto dim : shape.dimensions()) {
+    seed = HashCombine(seed, dim);
+  }
+  return HashCombine(seed, static_cast<int>(shape.element_type()));
+}
+
+}  // namespace
+
+hash_t ShapeHash(const Shape& shape) {
+  hash_t hash = 0xa5d2d6916;
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/third_party/computation_client/ltc_util.cc:23]: Yes, deterministic hash is very useful when debugging across multiple runs. It makes it possible to observe invariant properties and correlations between runs knowing for sure that a node is the same thanks to the hash. It's also useful for persistent caches, across multiple training runs.

File: lazy_tensor_core/lazy_tensors/shape.cc [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#include "lazy_tensors/shape.h"
+
+namespace lazy_tensors {
+
+void Shape::DeleteDimension(int64 dim_to_delete) {
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/lazy_tensors/shape.cc:5]: Completely removing the dimension at the given index. It reduces the rank by 1.

File: lazy_tensor_core/lazy_tensor_core/csrc/ops/amp_foreach_non_finite_check_and_unscale.cpp [hunk 3]
```diff
@@ -11,6 +11,18 @@ namespace ir {
 namespace ops {
 namespace {
 
+lazy_tensors::Shape NodeOutputShape(const OpList& inputs,
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/lazy_tensor_core/csrc/ops/amp_foreach_non_finite_check_and_unscale.cpp:14]: IIRC it marks whether during the unscale of gradients some of them become infinite. Core documentation isn't very useful either: https://pytorch.org/cppdocs/api/function_namespaceat_1aa03dbbeba05a53919db459b399e9a710.html, but the high level idea is described here: https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.GradScaler.step.

File: lazy_tensor_core/lazy_tensors/core/platform/hash.cc [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+uint64 DecodeFixed64(const char* ptr) {
+  // Load the raw bytes
+  uint64 result;
+  memcpy(&result, ptr, sizeof(result));  // gcc optimizes this to a plain load
+  return result;
+}
+
+}  // namespace
+
+uint64 Hash64(const char* data, size_t n, uint64 seed) {
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/lazy_tensors/core/platform/hash.cc:22]: This is taken from TF and makes it possible to compare the XLA plugin with PT/XLA, which was useful (and might still be) for debugging. It doesn't say what algorithm this is and git blame points to the initial commit 6 years ago.

We can either use MurmurHash since it doesn't need to be cryptographic or credit TF authors.

File: lazy_tensor_core/lazy_tensors/core/platform/hash.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+// The second type argument is only used for SFNIAE below.
+template <typename T, typename = void>
+struct hash {
+  size_t operator()(const T& t) const { return std::hash<T>()(t); }
+};
+
+template <typename T>
+struct hash<T, typename std::enable_if<std::is_enum<T>::value>::type> {
+  size_t operator()(T value) const {
+    // This works around a defect in the std::hash C++ spec that isn't fixed in
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/lazy_tensors/core/platform/hash.h:35]: Just the operator, but shouldn't be necessary anymore, I don't think we still build with gcc 4.8.

File: lazy_tensor_core/lazy_tensors/literal.cc [hunk 6]
```diff
@@ -13,4 +14,21 @@ Literal::Literal(const Shape& shape) : shape_(shape) {
 
 const Shape& Literal::shape() const { return shape_; }
 
+size_t Literal::Hash() const {
+  size_t hash_value = ShapeUtil::Hash(shape());
+
+  ShapeUtil::ForEachSubshape(
+      shape(), [&](const Shape& subshape, const ShapeIndex& index) {
+        if (!subshape.IsArray()) {
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/lazy_tensors/literal.cc:22]: It does, I thought this was going to be needed but I'm pretty sure literal tuples will never be a thing. I'll simplify this.

File: lazy_tensor_core/lazy_tensors/primitive_util.h [hunk 7]
```diff
@@ -126,6 +126,9 @@ inline bool IsIntegralType(PrimitiveType type) {
   return IsUnsignedIntegralType(type) || IsSignedIntegralType(type);
 }
 
+// Returns true if values of the given primitive type are held in array shapes.
+bool IsArrayType(PrimitiveType primitive_type);
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/lazy_tensors/primitive_util.h:130]: Is it still? I've checked and it looks ok to me, not hard coded in the current version.

The comment explains it, it returns true iff the primitive type can be the base type of an array - arrays can only hold true primitive types, not tuples. For example, the shape of a tensor is always an array, elements cannot be tuples.

File: lazy_tensor_core/lazy_tensors/shape.h [hunk 8]
```diff
@@ -50,7 +52,7 @@ class Shape {
   bool is_dynamic_dimension(int dimension) const { return false; }
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/lazy_tensors/shape.h:53]: I don't have test coverage for it yet, but I think it's reasonable since the setter is implemented. Will do it.

File: lazy_tensor_core/lazy_tensors/shape_util.cc [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+// Helper for ForEachSubshape which visits the subshapes of the given shape in
+// DFS pre-order starting with the index.
+Status ForEachSubshapeHelper(const Shape& shape,
+                             const ShapeUtil::StatusVisitorFunction& func,
+                             ShapeIndex* index) {
+  TF_RETURN_IF_ERROR(func(shape, *index));
+  if (shape.IsTuple()) {
+    for (int64 i = 0; i < ShapeUtil::TupleElementCount(shape); ++i) {
+      index->push_back(i);
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/lazy_tensors/shape_util.cc:24]: The `func` uses it to know where it is in a complex shape. This information could be used to decide what layout to be assigned to a specific sub-shape, for example. In practice, I don't think that information is truly used - maybe for pretty printing?

We can remove this and simplify the callers, I don't think the XLA plugin relies on it.

File: lazy_tensor_core/lazy_tensors/shape_util.cc [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+                                const VisitorFunction& func) {
+  ShapeIndex index;
+  ForEachSubshapeHelper(
+      shape,
+      [&func](const Shape& subshape, const ShapeIndex& index) {
+        func(subshape, index);
+        return Status::OK();
+      },
+      &index)
+      .IgnoreError();
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/lazy_tensors/shape_util.cc:45]: You can ignore the error because the lambda returns `Status::OK()`. This is inherited from TF, largely to keep code as similar as possible to PT/XLA and minimize chances of breakage. We can go more aggressively about simplifying these patterns now, if you'd like, at the cost of diverging a bit further from PT/XLA.

**Comment by asuhan** [comment]
@wconstab We do have unit tests for the basics, including the hash - have a look at test/cpp/test_ir.cpp. All functionality is covered by the tests in the XLA plugin, but adequate testing without any back-end is almost impossible. My thinking is that we'll have a first-party plugin (GPU / TorchScript) which can then be used for exhaustive testing.

A high-level observation: I've kept everything I extracted from PT/XLA as close to the original as possible, which in turn has some impact on the TF-like support code. That code is going to change, but I thought it'd be better to not do it just yet in the interest of minimizing this PR. We can chase those idiosyncrasies in this PR, if you prefer.

File: lazy_tensor_core/lazy_tensor_core/csrc/ops/get_dimensions_size.cpp [hunk 11]
```diff
@@ -17,10 +17,7 @@ GetDimensionsSize::GetDimensionsSize(
            lazy_tensors::ShapeUtil::MakeShape(
                GetShapeDimensionType(/*device=*/nullptr), {}),
            /*num_outputs=*/1, lazy_tensors::util::MHash(dimensions)),
-      dimensions_(std::move(dimensions)) {
-  SetShapeDeferred(
-      [&]() { return compiler::NodeLowering::Get()->Infer(this); });
-}
+      dimensions_(std::move(dimensions)) {}
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/lazy_tensor_core/csrc/ops/get_dimensions_size.cpp:20]: I wouldn't call it redundant, but it has the same purpose - it computes the output shape. The shape inference has to be deferred in the same manner as the lowering, since neither is provided by the core.

This is a stop-gap solution to this problem: https://github.com/pytorch/pytorch/blob/d18ba4b0305091b49be631558fc39127042b28f9/lazy_tensor_core/lazy_tensor_core/csrc/compiler/node_lowering.h#L21. It's a pretty high priority item to address, but I didn't have time for it yet.

File: lazy_tensor_core/lazy_tensors/core/platform/hash.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (wconstab) [reply lazy_tensor_core/lazy_tensors/core/platform/hash.cc:22]: ok, pointer to TF commit would be fine i think.  I support the idea of not changing behavior now and maintaining debug-ability.  If its a direct copy paste of something working 6 years and covered by backend tests then I'm ok with it.

**Review by wconstab — APPROVED** [review]
ok, your comments addressed my questions.  I did request a cosmetic change to add comments in one place, but otherwise lgtm as is and prefer the minimal change strategy.
